src/main/io/gps.c

```diff
@@ -573,110 +755,393 @@ static void ubloxSetMessageRate(uint8_t messageClass, uint8_t messageID, uint8_t
+
+                if (*cmd) {
+                    // more commands to send
+                    commandOffset = cmd - commands;
+                } else {
+                    gpsData.state_position++;
+                    commandOffset = 0;
+                }
+            } else
+#else
+            {
+                serialSetBaudRate (gpsPort, baudRates[gpsInitData[gpsData.baudrateIndex].baudrateIndex]);
+            }
+#endif
+            gpsSetState(GPS_STATE_RECEIVING_DATA);
+        break;
     }
-    ubloxSendConfigMessage(&tx_buffer, MSG_CFG_SBAS, sizeof(ubxCfgSbas_t));
 }
+#endif // USE_GPS_NMEA
+
+#ifdef USE_GPS_UBLOX
 
 void gpsInitUblox(void)
 {
-    uint32_t now;
-    // UBX will run at the serial port's baudrate, it shouldn't be "autodetected". So here we force it to that rate
-
     // Wait until GPS transmit buffer is empty
     if (!isSerialTransmitBufferEmpty(gpsPort))
         return;
 
     switch (gpsData.state) {
         case GPS_STATE_INITIALIZING:
-            now = millis();
-            if (now - gpsData.state_ts < GPS_BAUDRATE_CHANGE_DELAY)
-                return;
 
-            if (gpsData.state_position < GPS_INIT_ENTRIES) {
-                // try different speed to INIT
-                baudRate_e newBaudRateIndex = gpsInitData[gpsData.state_position].baudrateIndex;
+            // DEBUG_SET(DEBUG_GPS_CONNECTION, 7, initBaudRateCycleCount);
 
-                gpsData.state_ts = now;
+            if (initBaudRateCycleCount > BAUD_COUNT * 2) {
+                // Give up after 32 connection attempts ??
+                initBaudRateIndex = gpsInitData[gpsData.baudrateIndex].baudrateIndex;
+            }
 
-                if (lookupBaudRateIndex(serialGetBaudRate(gpsPort)) != newBaudRateIndex) {
-                    // change the rate if needed and wait a little
-                    serialSetBaudRate(gpsPort, baudRates[newBaudRateIndex]);
-#if DEBUG_SERIAL_BAUD
-                    debug[0] = baudRates[newBaudRateIndex] / 100;
-#endif
-                    return;
-                }
+            if (initBaudRateIndex < BAUD_COUNT) {
+                // initBaudRateIndex is initialised to BAUD_COUNT of 16, so this falls through at the start
+                // but if there has been a connection, it gets set to gpsData.state_position
+                // this performs a reconnection attempt at the last successful value if we reconnect
+                serialSetBaudRate(gpsPort, baudRates[initBaudRateIndex]);
+                serialPrint(gpsPort, gpsInitData[gpsData.baudrateIndex].ubx);
+                // should this be set to the same initBaudRateIndex value ??
+                gpsSetState(GPS_STATE_CHANGE_BAUD);
+                return;
+            }
+            uint32_t now = millis();
+
+            // send class message requests every 120ms to test the connection.
+            // cannot assume that the GPS update rate is 100hz, can be 1000hz if the buffer is full
+            static uint32_t delay = 0;
```

Comment:
Thanks, won't use micros() :-)
I am wondering about adding 'now' to the gpsData enum, setting it to `millis()` at the top of `gpsUpdate`, and then using `gpsData.now` whenever we need to know the current value of `millils()`.  
This may avoid recalculating `millis()`, many times, in functions called by `gpsUpdate()`.

Reply:
Don't worry about it now.
IMO the situation is not that bad.